File: sdk/python/tesela/builder/_pipeline.py
from __future__ import annotations

from typing import Any

from tesela.builder._common import _replace_by

class PipelineBuilder:
    def __init__(self, app: App, api_name: str):
        self._app = app
        self._data: dict[str, Any] = {"api_name": api_name, "steps": []}

    def step(self, api_name: str, source: str, target: str, *,
             expression: str = "", language: str = "",
             when: str = "", on_error: str | dict = "",
             dynamic_source: str = "", kind: str = "") -> "PipelineBuilder":
        s: dict[str, Any] = {"api_name": api_name, "source": source, "target": target}
        if expression:
            s["expression"] = expression
        if language:
            s["language"] = language
        if when:
            s["when"] = when
        if on_error:
            s["on_error"] = on_error if isinstance(on_error, dict) else {"strategy": on_error}
        if dynamic_source:
            s["dynamic_source"] = dynamic_source
        if kind:
            s["kind"] = kind
        self._data["steps"].append(s)
        return self

    def decision(self, api_name: str, expression: str) -> "PipelineBuilder":
        return self.step(api_name, "_none", "_none", expression=expression, kind="decision")

    def schedule(self, cron: str) -> "PipelineBuilder":
        self._data["schedule"] = {"Cron": cron} if cron != "manual" else "manual"
        return self

    def mode(self, m: str) -> "PipelineBuilder":
        self._data["mode"] = m
        return self

    def display(self, v: str) -> "PipelineBuilder":
        self._data["display"] = v
        return self

    def description(self, v: str) -> "PipelineBuilder":
        self._data["description"] = v
        return self

    def context(self, key: str, value: Any) -> "PipelineBuilder":
        self._data.setdefault("context", {})[key] = value
        return self

    def metadata(self, key: str, value: Any) -> "PipelineBuilder":
        self._data.setdefault("metadata", {})[key] = value
        return self

    def done(self) -> App:
        self._app._pipelines = _replace_by(self._app._pipelines, self._data)
        return self._app

File: crates/tesela-ir/src/pipeline.rs
//! Transform pipeline types.

use tesela_core::{ApiName, Value};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// A DAG of transform steps that produce object types from other object types.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TransformPipeline {
    /// API name.
    pub api_name: ApiName,
    /// Display name.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub display: Option<String>,
    /// Description.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub description: Option<String>,
    /// Ordered list of transform steps.
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub steps: Vec<TransformStep>,
    /// Cron-style execution schedule.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub schedule: Option<PipelineSchedule>,
    /// Execution mode.
    #[serde(default)]
    pub mode: ExecutionMode,
    /// Metadata.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub metadata: Option<BTreeMap<String, Value>>,
}

/// A single step in a transform pipeline.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TransformStep {
    /// API name (must be unique within the pipeline).
    pub api_name: ApiName,
    /// Source object type or other pipeline to read from.
    pub source: ApiName,
    /// Target object type to write into.
    pub target: ApiName,
    /// Expression that maps source records to target records.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub expression: Option<String>,
    /// Language for the expression ("cel", "sql", "identity").
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub language: Option<String>,
    /// Only process records modified after the last successful run.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub watermark_property: Option<ApiName>,
}

/// How the pipeline refreshes the target.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum ExecutionMode {
    /// Only process records changed since the last run.
    #[default]
    Incremental,
    /// Truncate and re-populate the target on every run.
    Snapshot,
}

/// When the pipeline should run automatically.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PipelineSchedule {
    /// Standard cron expression (e.g. `"0 * * * *"`).
    Cron(String),
    /// Only triggered explicitly via the API.
    Manual,
}

/// Result returned after executing a pipeline.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PipelineResult {
    /// Records written across all steps.
    pub records_written: i64,
    /// Wall-clock execution time in milliseconds.
    pub duration_ms: u64,
    /// Per-step error messages (non-fatal if the step is optional).
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub errors: Vec<String>,
}
